In [1]:
import pandas as pd

# loading the reddit_dataset
df = pd.read_csv(r'C:\Users\acer\Dropbox\PC\Downloads\reddit_opinion_ru_ua.csv')

# displaying the dataset
df                                    

comment_id  score                                          self_text  \
0          ku4hnmi      1  &gt; As far as the NATO expansion, you are rig...   
1          ku4hl9e      1  &gt;Please, if you can, send more\n\nOr even b...   
2          ku4hk9g      1  Someone should tell them you can bury more per...   
3          ku4hj68      1  I'd argue that a russian welder in Finland ear...   
4          ku4hj13      1  Yeah I've been living in France for years and ...   
...            ...    ...                                                ...   
1846441    cmqs3d9      1  Article is in Estonian, but here's my English ...   
1846442    cmf3xhd      3                                      I have iPhone   
1846443    cm4iugq      2             Merkel actually left at the same time.   
1846444    cm3dnw7      1      Lol that title. You cant deny an accusation.    
1846445    cm0ztol      2                                     I denies this.   

                     subreddit         created_time  post_id  \
0                   EndlessWar  2024-03-09 20:44:58  1b9m46z   
1          UkraineRussiaReport  2024-03-09 20:44:36  1bapdov   
2        UkraineWarVideoReport  2024-03-09 20:44:26  1banbxs   
3        UkraineWarVideoReport  2024-03-09 20:44:16  1baewv1   
4                       europe  2024-03-09 20:44:14  1bahtsh   
...                        ...                  ...      ...   
1846441           RussiaDenies  2014-12-10 14:36:31   2ov646   
1846442           RussiaDenies  2014-11-28 06:25:13   2n9eco   
1846443           RussiaDenies  2014-11-17 00:25:22   2mduol   
1846444           RussiaDenies  2014-11-15 18:28:20   2m33hb   
1846445           RussiaDenies  2014-11-13 02:18:53   2m43hd   

                author_name  controversiality  ups  downs  ...  \
0                  IntnsRed                 0    1      0  ...   
1        Current-Power-6452                 0    1      0  ...   
2             EDF_AirRaider                 0    1      0  ...   
3          Tyhjantoimittaja                 0    1      0  ...   
4                 Goldencol                 0    1      0  ...   
...                     ...               ...  ...    ...  ...   
1846441           Martenz05                 0    1      0  ...   
1846442      putin_vladimir                 0    3      0  ...   
1846443           [deleted]                 0    2      0  ...   
1846444           Player276                 0    1      0  ...   
1846445      putin_vladimir                 0    2      0  ...   

         user_link_karma user_comment_karma  user_total_karma  post_score  \
0               623553.0            56940.0          686397.0          22   
1                 5426.0            30156.0           35582.0          38   
2                    6.0             6140.0            6146.0         617   
3                 4833.0              838.0            5722.0        2794   
4                  416.0            23777.0           24462.0        1655   
...                  ...                ...               ...         ...   
1846441           4558.0            44942.0           49520.0           5   
1846442              0.0                0.0               0.0           6   
1846443              0.0                0.0               0.0          10   
1846444           3524.0            70879.0           75152.0           5   
1846445              0.0                0.0               0.0          19   

                                            post_self_text  \
0         I don't support either side in the war but i ...   
1        THESE are the monster Challenger 2 tanks blast...   
2                                                      NaN   
3                                                      NaN   
4                                                      NaN   
...                                                    ...   
1846441                                                NaN   
1846442                                                NaN   
18

Since the reddit_dataset is a collection of comments made on certain posts, cluster sampling could prove to be an efficient approach. In cluster sampling, the population is divided into clusters and then random samples of clusters are selected, containing all entries within the selected clusters.

In [2]:
# counting the number of unique post IDs
num_unique_post_ids = df['post_id'].nunique()

# displaying the number of unique post IDs
print("Number of Unique Post IDs:", num_unique_post_ids)

Number of Unique Post IDs: 58780


In [3]:
import numpy as np

# Setting random seed for reproducibility
np.random.seed(42)

# defining the number of clusters to sample
num_clusters_to_sample = 6500

# randomly selecting clusters to sample
clusters_to_sample = np.random.choice(df['post_id'].unique(), size=num_clusters_to_sample, replace=False)

# performing cluster sampling by selecting all data points belonging to the sampled clusters
sampled_df = df[df['post_id'].isin(clusters_to_sample)]

# displaying the sampled DataFrame
sampled_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 202942 entries, 4 to 1846422
Data columns (total 24 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   comment_id                  202942 non-null  object 
 1   score                       202942 non-null  int64  
 2   self_text                   202942 non-null  object 
 3   subreddit                   202942 non-null  object 
 4   created_time                202942 non-null  object 
 5   post_id                     202942 non-null  object 
 6   author_name                 202942 non-null  object 
 7   controversiality            202942 non-null  int64  
 8   ups                         202942 non-null  int64  
 9   downs                       202942 non-null  int64  
 10  user_is_verified            202942 non-null  bool   
 11  user_account_created_time   196318 non-null  object 
 12  user_awardee_karma          202935 non-null  float64
 13  user_awarder_

Figuring out the unique ids in post_id column is crucial to figure the number of clusters and define the number of clusters to be sampled. Hence, the resulting 110MB dataset contains 202,942 records, fulfilling the project requirements. 

In [4]:
# setting index=False to exclude the index from the output
# storing the sampled_dataset
sampled_df.to_csv('data.csv', index=False)

Stratified cluster sampling could be an alternative approach to sample the dataset. This technique is a combination of two sampling techniques: stratified sampling and cluster sampling. It is used when the population can be naturally divided into distinct subgroups/stratas (subreddit), and within each stratum, clusters (post_id) of individuals or items exist.

In [7]:
# figuring the count of subreddits in the reddit_dataset
num_unique_post_ids = df['subreddit'].nunique()
print(num_unique_post_ids)

47


In [8]:
# grouping the DataFrame by 'subreddit' and aggregating the unique post IDs
unique_posts_per_subreddit = df.groupby('subreddit')['post_id'].nunique()

# displaying the result
print(unique_posts_per_subreddit)

subreddit
ANormalDayInRussia        323
ArtForUkraine             274
AskARussian              2251
AskReddit                 216
CombatFootage            2155
CombatFootageUkraine      490
Conservative              368
EndlessWar               1163
FreedomofRussia           806
GTAorRussia               128
InvasionOfUkraine         290
LiveUkraine               128
LoveForUkraine           1547
Military                  424
MilitaryPorn              160
N_N_N                     491
NonCredibleDefense        713
RussiaDenies              526
RussiaReplacement         165
RussiaUkraineWar2022     2305
RussianWarFootage        2231
UkraineConflict          1859
UkraineInvasionVideos    1196
UkraineLosses             266
UkraineOSINT              142
UkrainePics               137
UkraineRussiaReport      6300
UkraineWarFootage         179
UkraineWarReports         537
UkraineWarRoom            684
UkraineWarVideoReport    6942
Ukraine_UA                 50
UkrainianConflict        7668


In [11]:
# defining the number of clusters to sample from each stratum
clusters_per_stratum = 50

# performing stratified cluster sampling
sampled_clusters = []
for stratum in df['subreddit'].unique():
    clusters_in_stratum = df[df['subreddit'] == stratum]['post_id'].unique()
    sampled_clusters.extend(np.random.choice(clusters_in_stratum, clusters_per_stratum, replace=False))

# Sampled data
strata_sampled_data = df[df['post_id'].isin(sampled_clusters)]

# displaying the dataset information
strata_sampled_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 55934 entries, 299 to 1846439
Data columns (total 24 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   comment_id                  55934 non-null  object 
 1   score                       55934 non-null  int64  
 2   self_text                   55934 non-null  object 
 3   subreddit                   55934 non-null  object 
 4   created_time                55934 non-null  object 
 5   post_id                     55934 non-null  object 
 6   author_name                 55934 non-null  object 
 7   controversiality            55934 non-null  int64  
 8   ups                         55934 non-null  int64  
 9   downs                       55934 non-null  int64  
 10  user_is_verified            55934 non-null  bool   
 11  user_account_created_time   53267 non-null  object 
 12  user_awardee_karma          55932 non-null  float64
 13  user_awarder_karma         

In [10]:
strata_sampled_data.to_csv('sc_data.csv', index=False)  # Set index=False to exclude the index from the output

Since the minimum number of posts under the subreddits is 50, the number of clusters to sample from each stratum is set to 50. This approach results in a 44.3MB dataset containing 55934 entries. Hence, not letting us fulfill the project requirements.

Therefore, the cluster sampling approach was taken to sample the dataset. 